src/java/org/jivesoftware/openfire/group/GroupManager.java

```diff
@@ -213,11 +207,20 @@ public void adminRemoved(Group group, Map params) {
                 
                 // Remove only the collection of groups the member belongs to.
                 String member = (String) params.get("admin");
-                if(member != null) {
-	                groupMetaCache.remove(member);
-                }
+                evictCachedUserForGroup(member);
             }
 
+            private void evictCachedUserForGroup(String user) {
+                if(user != null) {
+                    JID userJid = new JID(user);
+                    if (XMPPServer.getInstance().isLocal(userJid)) {
+                        String username = userJid.getNode();
+                        synchronized (username.intern()) {

```

Comment:
The code appears to lock, always, based on the cache-key that is accessed. The same lock is used when data is added for a particular username in ```getSharedGroups()```, so in that sense it is "correct" to do that here too.

I do share @dwd's concern of using an interned string - more so when the interned string is a very generic one, such as a username. If any other code (unrelated to this functionality), applies the same technique of locking on a username, we're in trouble. A quick and dirty improvement would be to apply some kind of namespace. Instead of locking on "username", we could lock on "classname+username" - or something similar.